iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/relocatable_ptr.hpp

```diff
@@ -0,0 +1,136 @@
+///       There should be no need for this, since as pass-around type
+///       regular pointers do the job just fine and do not incur
+///       the slight runtime overhead of a relocatable_ptr.
+///       There should be no memory overhead on 64 bit systems.
+template <typename T>
+class relocatable_ptr
+{
+  public:
+    /// @brief Construct from raw pointer.
+    relocatable_ptr(T* ptr = nullptr);
+    // {
+    //     m_offset = to_offset(ptr);
+    // }
+
+    /// @brief Construct from other relocatable pointer.
+    relocatable_ptr(const relocatable_ptr& other);
+
+    /// @brief Move construct from other relocatable pointer.
+    relocatable_ptr(relocatable_ptr&& other);
+
+    /// @brief Assign from relocatable pointer rhs.
+    relocatable_ptr& operator=(const relocatable_ptr& rhs);
+
+    /// @brief Move assign from relocatable pointer rhs.
+    relocatable_ptr& operator=(relocatable_ptr&& rhs);
+
+    /// @brief Get the corresponding raw pointer.
+    /// @return corresponding raw pointer
+    T* get();
+
+    /// @brief Get the corresponding raw pointer from const relocatable_ptr
+    /// @return corresponding raw pointer
+    const T* get() const;
+
+    /// @brief Dereference a relocatable_ptr.
+    /// @return reference to the pointee
+    /// @note not available for T=void
+    template <typename S = T>
+    S& operator*();
+
+    /// @brief Dereference a const relocatable_ptr.
+    /// @return reference to the pointee
+    /// @note not available for T=void
+    template <typename S = T>
+    const S& operator*() const;
+
+    /// @brief Get the corresponding raw pointer with arrow operator syntax.
+    /// @return corresponding raw pointer
+    T* operator->();
+
+    /// @brief Get the corresponding raw pointer with arrow operator syntax
+    ///        from a const relocatable_ptr.
+    /// @return corresponding raw pointer
+    const T* operator->() const;
+
+    /// @brief Convert to the corresponding raw pointer.
+    /// @return corresponding raw pointer
+    operator T*();
+
+    /// @brief Convert to the corresponding const raw pointer.
+    /// @return corresponding const raw pointer
+    operator const T*() const;
+
+  private:
+    using offset_t = uint64_t;
+
+    // This is safe since it is equivalent to point to the relocatable pointer
+    // second byte of the relocatable pointer itself which we define to be illegal.
+    // (it is no reasonable use-case)
+    // Note that 0 is equivalent to point to the relocatable pointer itself (i.e. this).
+    static constexpr offset_t NULL_POINTER_OFFSET = 1;
+
+    offset_t m_offset;
+
+    offset_t self() const;
+
+    offset_t to_offset(const void* ptr) const;
+
+    T* from_offset(offset_t offset) const;
+};
+
+/// @brief Compare relocatable_ptr with respect to logical equality.
+/// @return true if rhs and lhs point to the same location, false otherwise
+template <typename T>
+bool operator==(const relocatable_ptr<T>& lhs, const relocatable_ptr<T>& rhs);
```

Comment:
Why do you implement this as a free function instead of a member method?

Reply:
It is good practice to do so with these operators I think and necessary if one wants to have asymmetric operators: compare an S with a T and a T with S. Not directly the case here, or more precisely happens due to conversion: we can compare raw pointers with relocatable pointers. First I provided specific overloads for that (which needed to be non-member for the above reasons) but this causes issues with type deduction due to the implicit conversion (which I want to keep).

This is the historical reason. Another reason is basically summarized here https://embeddedartistry.com/fieldatlas/how-non-member-functions-improve-encapsulation/

There is also something like this in the usual books, but I did not look it up.